net: lib: threaded download_client and API updates

lemrey opened the pull request:
- Reworked library to use threads
- Rewrote buffer handling and removed separate request buffer

Futher, the API has been updated to fix two issues:

- The application was required to peek into the library's
  supposedly opaque instance data to poll the library state
- The library internal state was exposed to the application,
  while also being partially reported via events, thus being
  rather confusing.

Added functionality:
- Download from an offset
- Suspend/resume

lemrey commented:
@sigvartmh please verify if the functionality of the application is still intact.

sigvartmh commented:
Will do :)

NordicBuilder commented:
Found the following issues, please fix and resubmit:

## checkpatch issues

```
No additional types will be considered - file 'scripts/checkpatch/typedefsfile': No such file or directory
-:192: WARNING:FUNCTION_ARGUMENTS: function definition argument 'struct download_client *' should also have an identifier name
#192: FILE: include/download_client.h:171:
+void download_client_pause(struct download_client *);

-:193: WARNING:FUNCTION_ARGUMENTS: function definition argument 'struct download_client *' should also have an identifier name
#193: FILE: include/download_client.h:172:
+void download_client_resume(struct download_client *);

-:682: ERROR:C99_COMMENTS: do not use C99 // comments
#682: FILE: subsys/net/lib/download_client/src/http_download_client.c:260:
+			// thread dead?

- total: 1 errors, 2 warnings, 1005 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

Your patch has style problems, please review.

NOTE: Ignored message types: AVOID_EXTERNS BRACES CONFIG_EXPERIMENTAL CONST_STRUCT DATE_TIME FILE_PATH_CHANGES MINMAX MULTISTATEMENT_MACRO_USE_DO_WHILE NETWORKING_BLOCK_COMMENT_STYLE PRINTK_WITHOUT_KERN_LEVEL SPLIT_STRING VOLATILE

NOTE: If any of the errors are false positives, please report
      them to the maintainers.



```

osaether commented:
I tried _http_application_update_ and it hangs after printing the first two lines. This is what I see in the terminal:
`[00:03:19.132,934] <inf> download_client: Socket ready
[00:03:19.138,641] <inf> download_client: Downloading: /nordic-firm`

lemrey commented on subsys/net/lib/download_client/src/http_download_client.c:
```diff
@@ -7,21 +7,24 @@
+#include <misc/util.h>
 #include <logging/log.h>
 
-LOG_MODULE_REGISTER(http_dfu);
```
I agree, am planning to add that.

lemrey replied:
Added separate logging config.

lemrey commented on subsys/net/lib/download_client/src/http_download_client.c:
It was being handled in the previous implementation, I have not incurred in it.
If it can be removed I'll happily do that, who could confirm?

rlubos replied:
I'm pretty sure we got this fixed, @gbakke can you confirm?

gbakke replied:
That's correct @rlubos. Using the bsdlibrary 0.3.0+,the recv() with 0 flag should now block and not report EAGAIN anymore. Also, ETIMEDOUT should not be a concern as SO_RCVTIMEO is not set on the socket.

lemrey replied:
Removed `-EAGAIN`.

lemrey commented on include/download_client.h:
```diff
@@ -107,43 +80,31 @@ struct download_client {
+	size_t progress;
+	size_t offset;
+
+	int status;
```
`status` is leftover, will be removed.

lemrey replied:
Removed.

lemrey commented on include/download_client.h:
```diff
@@ -107,43 +80,31 @@ struct download_client {
-	/** Status of the transfer (see @ref download_client_status). */
-	volatile int status;
+
+	K_THREAD_STACK_MEMBER(thread_stack, 2048);
```
Perhaps could be made configurable.

lemrey replied:
Made configurable.

lemrey commented on include/download_client.h:
```diff
@@ -107,43 +80,31 @@ struct download_client {
+	K_THREAD_STACK_MEMBER(thread_stack, 2048);
+	struct k_thread thread;
+	k_tid_t tid;
+	struct k_sem sem;
```
`sem` is leftover and will be removed.

lemrey replied:
Removed.

lemrey commented on samples/nrf9160/http_application_update/src/main.c:
```diff
@@ -131,19 +126,18 @@ static int flash_page_erase_if_needed(u32_t address)
+	switch (event->id) {
+	case DOWNLOAD_CLIENT_EVT_FRAGMENT: {
+#if 0
+		if (dfu->file_size > PM_MCUBOOT_SECONDARY_SIZE) {
```
I think the best way to return the file size would be via an event, and the simplest way could be to add it to an existing event instead of sending out a separate one, just to keep the state machine simple.

lemrey replied:
Added a dedicated API.

osaether commented:
> @sigvartmh please verify if the functionality of the application is still intact, I only made it compile.

http_application_update works after removing the busy waiting in main. I will make a separate PR for it.

osaether commented:
> I tried _http_application_update_ and it hangs after printing the first two lines. This is what I see in the terminal:
> 
> ```
> [00:03:19.132,934] <inf> download_client: Socket ready
> [00:03:19.138,641] <inf> download_client: Downloading: /nordic-firm
> ```
It works after removing the busy waiting in main.

lemrey commented:
A tricky part is what to do when the download needs to be restarted; pausing/resuming the download can easily be done via `k_thread_*` APIs, but restarting the download is a bit tricky because a thread can't be re-spawned after it has returned AFAIK. @rlubos any ideas how to restart the download?

rlubos commented:
@lemrey If it's indeed not possible to restart the thread, then I guess we should not terminate it :) Perhaps a semaphore to block the thread execution before the loop would help here? In case of some kind of failure, we could jump back to the beginning of the thread.

And you'd probably need some flag to indicate that download is in progress, to be clear whether to resume the thread or feed the semaphore.

lemrey commented:
@rlubos do you think it's a good idea to make this library use a thread internally,
instead of letting the application decide whether they want to run it in a thread?

rlubos commented:
@lemrey I don't have strong opinion here. As typical, it's a the choice between ease of use vs felxibility. IMO having a separate thread to push DFU forward is a good enough approach.

osaether commented:
@lemrey Can you add a function for getting the size of the file to be downloaded?

lemrey commented:
@osaether yes, I will make the size available either via API directly or as a parameter in a event (where we are sure it has been retrieved).

lemrey commented on subsys/net/lib/download_client/src/http_download_client.c:
For now, we can let the application cleanup.
I should have mentioned that in the event documentation.

lemrey commented on subsys/net/lib/download_client/src/http_download_client.c:
Fixed, returning `-errno`.

lemrey commented:
- Cleaned up header
- Thread can now be "restarted" after socket errors, and when callback returns non-zero
- Made thread stack size configurable
- Added separate logging level for the library
- Added static assert on fragment/response buffer sizes
- Added function to retrieve file size

hakonfam commented on include/download_client.h:
```diff
@@ -17,220 +17,184 @@
  */
-typedef int (*download_client_event_handler_t)(struct download_client *client,
-				enum download_client_evt event, u32_t status);
+typedef int (*download_client_callback_t)(
```
Is it needed to have this return an int? AFAICS its return value is not used.

lemrey replied:
If non-zero is returned, the download will stop.

hakonfam commented on include/download_client.h:
```diff
@@ -17,220 +17,184 @@
+	bool has_header;
+
+	/** Server hosting the file, NULL terminated */
+	char *host;
```
Is there any reason why this and `file` is not `const` ?

lemrey replied:
That would limit one instance of the library to one server and one file, but in practice that limitation is artificial since the implementation can connect to a new server and download a new file. By letting both variables be modifiable, you can re-use one instance to download different files from different servers.

rlubos commented on samples/nrf9160/http_application_update/src/main.c:
```diff
@@ -52,46 +47,48 @@ void bsd_irrecoverable_error_handler(uint32_t err)
+	if (retval != 0) {
+		printk("download_client_connect() failed, err %d",
+			retval);
+			return 1;
```
nit: Indent

lemrey commented:
Only documentation left to review, I have added @ru-fu and @b-gent.

hakonfam commented:
IMO docs can be done in a second review, and is not affected by code freeze, I propose this goes in as soon as the code is accepted.

lemrey commented:
@b-gent thank you, I have squashed your commit.
